File: shared/public/ui/amount.js
/**
 * Reading and stepping an amount, DOM-free so it can be tested without one.
 *
 * The counterpart of `ui/time.js`: the rules live here and `ui/count-field.js`
 * is the control over them, exactly as `ui/reminder-field.js` sits over the
 * time rules. That split is what makes these testable at all.
 *
 * Why a parser rather than `<input type="number">`, which is what both places
 * that record an amount used to be. Measured in Chrome, typing into one with
 * the day editor's own attributes:
 *
 *   typed "8,5"   -> .value "85"     eight and a half recorded as eighty-five
 *   typed "abc"   -> .value ""       read as "no entry" — the day is DELETED
 *   typed "1e3"   -> .value "1e3"    Number() reads 1000
 *
 * The browser filters the keystrokes it does not like and hands back whatever
 * survived, so an unparseable amount is never an error to report — it is a
 * different amount, or nothing. The decimal comma is the one that matters: it
 * is what most of Europe's keyboards invite, `inputmode="decimal"` shows it,
 * and dropping it multiplies the answer by ten. `HabitFormScreen.parseAmount`
 * on the phone has a comment about the same input — and note it is NOT a mirror
 * of this and does not try to be: it reads a habit's TARGET where this reads a
 * DAY's amount. Be exact about where they part, because "not a mirror" is easy
 * to say too strongly. They agree about the FORM: the same digits-and-at-most-
 * one-separator shape, the same refusal of a sign, an exponent or hex, and —
 * on a `point` account, which is every account until somebody says otherwise —
 * the same thousands refusal since #111. They part in two places. The DOMAIN:
 * this one bounds and quantises to what `formatAmount` can show back, so
 * `3.14159265` is 3.141593 here and itself there. And the CONVENTION: `10.000`
 * is ten thousand here on a `comma` account and ten on the phone, always, which
 * is #108's remaining half and is written down as such rather than fixed by
 * copying this file a sixth time.
 *
 * The web's own target box is the gap that leaves. `index.html`'s
 * `target_value` is still an `<input type="number">` — the very control this
 * comment opens by measuring — so `8,5` typed into a habit's goal is still 85.
 * That is issue #156 and not this module's to fix from here.
 *
 * **The server imports this**, which is the one thing here with a cost. A
 * Discord modal is a box somebody types an amount into, arriving over a socket
 * rather than off a keyboard, and `shared/src/discord.js` had its own reading
 * of one until it read "10,000" as ten. Nothing in `shared/public` may import
 * from `shared/src` — the browser cannot see it — but node can read this, so
 * the rule is imported rather than mirrored. What that costs is the first line
 * of this file: "DOM-free" was a convenience for the tests and is now a
 * CONTRACT with a server. A `document` reached for anywhere in here takes the
 * reminder path down with it.
 */

/**
 * The amount in a box, or what is wrong with it.
 *
 * Three answers, and callers must tell them apart with `===` because two of
 * them are falsy — the same convention (and the same trap) as `parseTimeInput`:
 *
 *   `''`    the box is empty. That is "nothing is known about this day", which
 *           for a caller is a DELETE. It is NOT the amount zero.
 *   `null`  something was typed and it is not an amount. Say so; write nothing.
 *   number  a non-negative amount. `0` is a real answer — a stated lapse — and
 *           is why the two above cannot be collapsed into one falsy check.
 *
 * **Either separator is read as a decimal point, and a three-digit GROUP is
 * refused.** Those are one rule seen from two ends, and the second is the only
 * place the convention matters at all: a group is exactly three digits, so
 * `8,5` and `8.5` are eight and a half under both conventions and need nothing
 * decided about them. `10,000` and `10.000` are where the two readings part,
 * and which of them is the group is what `format` says.
 *
 * A form that mixes both — `1,000.5` — is refused rather than guessed at, and
 * so is a group under the convention in force. Refusing is loud; guessing is
 * silent and wrong by a factor of a thousand, which is how "10,000 steps" —
 * this module's own example habit — once became ten.
 *
 * The regex is deliberately stricter than `Number()`, which the root CLAUDE.md
 * already records as too generous about form: it reads `0x10` as 16 and `1e3`
 * as 1000. Neither is a thing anyone types into a box asking how many glasses
 * of water they drank.
 *
 * @param {string} raw
 * @param {'point'|'comma'} [format] which character the DECIMAL point is, from
 *   `resolveNumberFormat`. Defaults to the app's own, so a caller that has no
 *   account to ask behaves as this always did.
 * @returns {'' | null | number}
 */
export function parseAmount(raw, format = 'point') {
  const text = String(raw ?? '').trim();
  if (text === '') return '';

  // The thousands group, refused rather than read. Under `point` that is
  // `10,000` — the case that has been refused since #60 — and under `comma` it
  // is `10.000`, which is #108: ten to this parser and ten thousand to the
  // de-DE reader who typed it, silently, with nothing on screen to say which
  // was stored.
  //
  // A non-zero integer part is required, so it fires on a real group and not on
  // `0,255` or `,255`, where no thousands can precede the separator. Three
  // digits is what makes it a group; `8,5` and `0,25` were never ambiguous.
  //
  // Note what is NOT done here: a group is never ACCEPTED, even under a stated
  // convention that makes it unambiguous. `10,000` on a `point` account could
  // be read as ten thousand and is not, because the answer would then be wrong
  // by a thousand for anyone whose convention was resolved from a device rather
  // than chosen — the `auto` default, which is most people. Refusing costs one
  // sentence saying what to type; accepting costs a wrong row nobody sees.
  if (GROUPS[format === 'comma' ? 'comma' : 'point'].test(text)) return null;

  const decimal = text.replace(/,/g, '.');
  if (!/^(\d+(\.\d*)?|\.\d+)$/.test(decimal)) return null;

  const value = Number(decimal);
  if (!Number.isFinite(value) || value < 0) return null;

  // Bounded at both ends so that everything this accepts is something
  // `formatAmount` can show back faithfully. Without it the two disagree at the
  // extremes and the control ends up displaying a value its own `value()`
  // refuses: `String()` goes exponential at 1e21 and this parser rejects an
  // exponent, and anything under half a millionth formats to "0" — which the
  // next Save would rewrite into a stated lapse. A trillion is past any amount
  // a habit records.
  if (value > MAX_AMOUNT) return null;
  if (value > 0 && value < MIN_AMOUNT) return null;

  // Quantised to what `formatAmount` can show. Without this the domain is
  // shared only at the ends: `3.14159265` was accepted whole, displayed as
  // `3.141593` on the next open, and saved as that — a value changing behind a
  // Save nobody meant to change anything with. Six places is far beyond what a
  // habit records.
  return Number(value.toFixed(6));
}

/**
 * The range an amount may take, chosen so `parseAmount` and `formatAmount`
 * have the same domain — see the bounds check above for what goes wrong when
 * they do not.
 */
export const MAX_AMOUNT = 1e12;
export const MIN_AMOUNT = 1e-6;

/**
 * A thousands group, one pattern per convention: "10,000" and "1,500" where the
 * decimal point is a point, "10.000" and "1.500" where it is a comma. Not
 * "0,255" or ",255" under either — see `parseAmount`, which is the only place
 * the reasoning is written down.
 *
 * **Anchored, because a SERVER reads this now.** The Kotlin mirror and this
 * file's own first version were `[1-9]\d*,\d{3}(?!\d)` unanchored, which is
 * quadratic on a string of digits with no comma in it: every start position
 * scans to the end looking for one. Measured here — 2.6ms at 2,000 characters,
 * 42ms at 8,000, 655ms at 32,000 — which was nobody's problem while the only
 * caller was a keystroke in a text box, and became one the moment
 * `src/discord.js` began handing it a modal field off a socket. CodeQL called
 * it (js/polynomial-redos) on precisely that new path.
 *
 * The anchors are not a narrowing. Everything the loose form caught that this
 * does not — "12,345,678", "10,000.5", "1,500 steps" — has more than one
 * separator or a non-digit in it, so the decimal test below refuses it one line
 * later and `parseAmount` answers null either way. Checked rather than argued:
 * 38,416 concatenations of digit and separator fragments, zero disagreements.
 * The `0*` is the one clause that is load bearing rather than cosmetic: without
 * it "01,000" is not a group here, and "01.000" IS a number to the decimal
 * test, so a leading zero would have bought a silent 1.
 *
 * ONE declaration per convention, read by the parser and by nothing else —
 * `amountComplaint` asks the parser instead of copying either. Written twice it
 * drifted: dropping the `(?!\d)` from the complaint's copy alone offered
 * separator advice about "10,0000", which the parser reads as 10.
 */
const GROUPS = {
  point: /^0*[1-9]\d*,\d{3}$/,
  comma: /^0*[1-9]\d*\.\d{3}$/,
};

/**
 * Which character the decimal point is, in three tiers.
 *
 * `resolveTimeZone`'s shape and for its reason: the account's stated answer
 * wins, else what the device reports, else the app's own. `'auto'` is a STORED
 * value and not the absence of one — the distinction `theme: 'system'` and
 * `at_most_unlogged: 'default'` both already draw — so a device fact never
 * becomes a stored decision, and a de-DE reader on an en-US work laptop has
 * something to change.
 *
 * The device's report is a separator character rather than a locale, because
 * that is the only part of a locale this asks about and a caller with no device
 * to ask (the server, answering a Discord modal) then has nothing to invent.
 * The third tier is the app's own convention and NOT the process's locale: a
 * container's ICU default is not a fact about the person typing, which is the
 * same trap `deliverAccount` fell into over `notifyTimezone`.
 *
 * @param {string} [setting] the account's `numberFormat`
 * @param {string} [deviceSeparator] what the device writes, from
 *   `deviceDecimalSeparator()` in a browser and nothing at all on a server
 * @returns {'point'|'comma'}
 */
export function resolveNumberFormat(setting, deviceSeparator) {
  if (setting === 'point' || setting === 'comma') return setting;
  if (deviceSeparator === ',') return 'comma';
  return 'point';
}

/**
 * What THIS device writes a decimal point as, for `resolveNumberFormat`'s
 * middle tier.
 *
 * An observation, never stored — the `device_clock` distinction one file over.
 * `formatToParts` rather than a locale string because the answer wanted is one
 * character and locale tags do not carry it (`de-AT` and `de-CH` disagree).
 * Anything that is not a comma resolves to `point`, which covers the Arabic and
 * Persian separators this app has no convention for.
 *
 * **The server must not call this.** In node it answers from the container's
 * ICU default, which is a fact about the host and not about the person typing;
 * `src/discord.js` passes no device and takes the third tier.
 *
 * @returns {string}
 */
export function deviceDecimalSeparator() {
  try {
    return new Intl.NumberFormat().formatToParts(1.5)
      .find((part) => part.type === 'decimal')?.value ?? '';
  } catch {
    return '';
  }
}

/**
 * Why `parseAmount` refused, for somebody looking at the box.
 *
 * "Not an amount" is true of `eight` and unhelpful for `10,000`, which IS a
 * number and a perfectly reasonable thing to type — it is refused because it is
 * AMBIGUOUS, not because it is nonsense, and a refusal the user cannot act on is
 * only half better than the silent ten it replaced. They typed their step goal
 * the way their phone's keyboard and their own country write it, and were told
 * it was not a number.
 *
 * The ADVICE agrees with the Kotlin `amountComplaint`'s — both say to type it
 * without the thousands separator, and about `10,000`, which is the input the
 * test uses, both name 10000 — and a test reads that string out of
 * `HabitFormScreen.kt` rather than trusting this comment, because two clients
 * telling somebody to type different things about the same input is a
 * difference with nothing behind it. It is agreement and not identity: the
 * phone has no room to quote what was typed and this box does, so the sentences
 * differ and the test checks the actionable core rather than the wording.
 *
 * Note the two parsers are NOT mirrors — see this module's header — so what is
 * pinned here is narrower than it looks: they happen to answer the same way
 * about a thousands group, and this makes them SAY the same thing about it.
 * Nothing about the rest of their domains is claimed.
 *
 * The predicate is `parseAmount` itself, twice, and not a copy of anything: the
 * box was refused, and taking the commas out makes it an amount. That is what
 * the advice CLAIMS, so it is the honest thing to test — and it cannot fire on
 * a case the parser accepts, which would tell somebody to change an entry that
 * was going to be stored correctly. It replaced a second copy of the parser's
 * thousands regex, which had already drifted from it once.
 *
 * **What is quoted must be true of what was quoted.** The first version of this
 * named the two readings — *could be ten thousand or ten and a half* — beside
 * whatever the user had typed, so "1,500" was told it might be ten thousand.
 * That is the module's own example of the ambiguity (`parseAmount` says
 * fifteen hundred or one and a half, correctly, forty lines up) with the wrong
 * number attached to it. Naming the reading is what made it specific and is
 * what made it false, so the sentence names the AMBIGUITY, which is the same
 * whatever the digits are, and the ADVICE carries the specifics instead.
 *
 * That advice is the user's own number with the commas taken out, because
 * "like 10000" is an example where "like 1500" is an instruction. Which makes
 * the suggestion something that can itself be wrong, so it is RUN THROUGH THE
 * PARSER before it is offered — and that decides the whole branch, not just the
 * number in it: `1,500 steps` holds a thousands group and is not ambiguous, it
 * is not an amount, and telling somebody to drop a comma that is not their
 * problem sends them round the loop again. A box may not suggest something it
 * would refuse. So the group is necessary and the suggestion working is what
 * makes it sufficient.
 *
 * The character it names is the one the convention in force groups with, so the
 * sentence is about what the reader is actually looking at. "Ambiguous" is
 * still the right word under a STATED convention, which is the one case where
 * the input is not: it is the notation that is ambiguous, and refusing a group
 * outright is what keeps the answer from depending on how it was resolved — see
 * `parseAmount`, which declines to accept one for exactly that reason.
 *
 * @param {string} raw exactly what is in the box
 * @param {'point'|'comma'} [format] as `parseAmount`
 * @returns {string} a sentence to show
 */
export function amountComplaint(raw, format = 'point') {
  const text = String(raw ?? '').trim();
  const grouper = format === 'comma' ? '.' : ',';
  const plain = text.split(grouper).join('');
  return parseAmount(text, format) === null
    && typeof parseAmount(plain, format) === 'number'
    ? `"${text}" is ambiguous — a ${format === 'comma' ? 'dot' : 'comma'} can `
      + `separate thousands or stand for a decimal point. Type it without the `
      + `thousands separator, like ${plain}.`
    : `"${text}" is not an amount — type a number like 8 or 8.5.`;
}

/**
 * How much one press of − or + should move, for a habit with this target.
 *
 * A spinner that steps by 1 is right for "8 glasses" and useless for "10,000
 * steps" — which is the complaint `<input type="number">`'s arrows earned. So
 * the step comes from the goal: about an eighth of it, snapped to a number a
 * person would have chosen.
 *
 * Snapped rather than computed, because `target / 8` is 1250 for 10,000 steps
 * and nobody counts in 1250s. The ladder is the set of round numbers, and the
 * answer is the largest one that does not overshoot an eighth of the target.
 *
 * A habit with no target — 0, which `parseHabit` accepts — gets 1. There is
 * nothing to derive from, and 1 is the step the field had before any of this.
 *
 * @param {number} target
 * @returns {number}
 */
export function stepFor(target) {
  const goal = Number(target);
  if (!Number.isFinite(goal) || goal <= 0) return 1;

  const ladder = [0.1, 0.25, 0.5, 1, 2, 5, 10, 25, 50, 100, 250, 500, 1000, 2500, 5000];
  // A whole-number goal gets a whole-number step. Below a target of 8 an eighth
  // is a fraction, so "3 glasses" stepped to 3.25 glasses — an amount in a unit
  // that does not divide. The goal's own form is the honest signal for that:
  // half a kilometre is a real thing to record, a quarter of a glass is not.
  const floor = Number.isInteger(goal) ? 1 : ladder[0];
  const wanted = goal / 8;

  let step = floor;
  for (const rung of ladder) if (rung <= wanted && rung >= floor) step = rung;
  return step;
}

/**
 * An amount as a person would write it: `3`, not `3.0`.
 *
 * Floating point is why this is not `String(n)` — stepping 0.1 up from 0.2
 * lands on 0.30000000000000004, and a box that shows that has stopped being a
 * count. Rounded to a sane number of places and then trimmed, which also makes
 * the arrows idempotent: pressing + then − returns the value it started on.
 *
 * It takes the convention as well, because a box that reads `8,5` and then
 * writes `8.5` back into itself has told its owner they typed it wrong. Nothing
 * is grouped at any size — `10000` and never `10,000` — which is what keeps
 * this inside `parseAmount`'s domain rather than producing the one form it
 * refuses. There is no thousands separator to get right here, only a decimal
 * point.
 *
 * Note the SERVER never renders through this: `answerText` in notify.js writes
 * an amount into a Discord message with `String()`, and giving that the
 * account's convention is a separate decision about prose in a channel other
 * people can read.
 *
 * @param {number} n
 * @param {'point'|'comma'} [format] as `parseAmount`
 * @returns {string}
 */
export function formatAmount(n, format = 'point') {
  const value = Number(n);
  if (!Number.isFinite(value)) return '';
  const shown = Number(value.toFixed(6));

  // Never render a real amount as nothing. `parseAmount` bounds what can be
  // TYPED, and that is not a bound on what can arrive: neither `parseEntry`
  // nor the habit's target is bounded server-side, so an import, the phone or
  // an older client can store a value smaller than this can show. As "0" it
  // would be rewritten into a stated lapse by the next Save; as its raw self
  // it is refused by `parseAmount`, and being refused is loud.
  const text = shown === 0 && value !== 0 ? String(value) : String(shown);
  return format === 'comma' ? text.replace('.', ',') : text;
}

/**
 * Move an amount by one step, never below zero.
 *
 * The floor is not a validation rule borrowed from the server — it is what the
 * control can offer. An amount is non-negative (`parseEntry` refuses the rest),
 * so a − that walks into a value the server would reject is a button that
 * produces an error, and zero is a real answer to stop at: a stated lapse.
 *
 * @param {number|''|null} current what `parseAmount` said about the box
 * @param {number} step
 * @param {1|-1} direction
 * @param {number} target used only when the box holds nothing to step from
 * @returns {number}
 */
export function stepAmount(current, step, direction, target = 0) {
  // Nothing usable in the box: the first press offers the goal rather than
  // stepping from zero, since that is nearly always what is about to be
  // recorded — the same reason the phone's count dialog prefills the target.
  if (typeof current !== 'number') {
    const goal = Number(target);
    if (direction > 0 && Number.isFinite(goal) && goal > 0) return goal;
    return direction > 0 ? step : 0;
  }
  return Math.max(0, Number((current + step * direction).toFixed(6)));
}

File: android-native/app/src/main/java/com/habiterall/app/ui/Amount.kt
package com.habiterall.app.ui

import java.math.BigDecimal
import java.text.DecimalFormatSymbols
import java.util.Locale

/**
 * Which way a decimal point is spelled — `.` or `,`.
 *
 * There is exactly one reader of a typed amount now ([parseAmount]), used by
 * the habit form, the notification number pad and the day dialog's count
 * field alike; this is what tells all three which spelling of a thousands
 * group to refuse. See [deviceAmountFormat] for where the value comes from
 * and `android-native/CLAUDE.md` for why this is deliberately NOT a sixth
 * hand-written mirror of a server setting.
 */
enum class AmountFormat { POINT, COMMA }

/**
 * What THIS device writes a decimal point as. Asked at PARSE time, never held.
 *
 * `Locale.getDefault()` can change while the process is running — a settings
 * change, or the OS itself — so this is a function and never a `val` at file
 * scope, the same rule `ui/count-field.js` states for the web about the
 * locale or the setting changing while the module is loaded. Only a literal
 * `,` resolves to [AmountFormat.COMMA]; anything else — including the Arabic
 * decimal separator `٫`, which this app has no convention for — is
 * [AmountFormat.POINT].
 */
internal fun deviceAmountFormat(): AmountFormat =
    if (DecimalFormatSymbols.getInstance(Locale.getDefault()).decimalSeparator == ',')
        AmountFormat.COMMA else AmountFormat.POINT

/**
 * An amount as a person would write it: `3`, not `3.0` — and spelled the way
 * [format] reads a decimal point back, or a box that reads "8,5" and then
 * writes "8.5" back into itself has told its owner they typed it wrong: what
 * this writes goes straight back into [parseAmount] on the next Save, in the
 * habit form, the day dialog's [CountDialog] and the notification number pad
 * alike. See `formatAmount`'s own doc comment in `shared/public/ui/amount.js`
 * for the same argument made about the web.
 *
 * **Grouping stays at no size under either convention** — `10000`, never
 * `10.000` or `10,000` — which is what keeps the output inside
 * [parseAmount]'s own domain rather than producing the one form it refuses.
 * There is no thousands separator to get right here, only which character is
 * the decimal point.
 *
 * **And it is a PLAIN string, never Java's exponent form, which is the second
 * half of staying inside that domain and was missed the first time.**
 * `Double.toString` switches to scientific notation below 1e-3 and at or above
 * 1e7, so it wrote `5.0E-4` for an amount of 0.0005 and `1.23456785E7` for one
 * of 12345678.5 — and [DECIMAL] admits no exponent, so the box was prefilled
 * with a string its own parser refused. Same shape as the comma bug beside it
 * and reachable on the same three surfaces: a habit whose target is under a
 * thousandth opened with the whole form unsaveable, and (once these two dialogs
 * started going through [parseAmount] at all) a DAY whose recorded amount is
 * under a thousandth opened with Save dead over the number the user themselves
 * recorded. Both values are ones [parseAmount] accepts as TYPED input, so this
 * is a formatter that could not show back what its own parser had just taken.
 *
 * `BigDecimal.valueOf` takes `Double.toString`'s canonical shortest digits and
 * `toPlainString` spells them without an exponent; `stripTrailingZeros` plus the
 * `scale() <= 0` branch is what keeps `3.0` reading as `3` rather than `3.0` or
 * `3E+1`-style scientific for a large integral. Nothing here bounds the domain
 * the way `shared/public/ui/amount.js` does — that file quantises to six places
 * and refuses past 1e12, and the two deliberately part there — this only makes
 * sure that whatever IS shown can be read back.
 */
internal fun formatAmount(v: Double, format: AmountFormat = deviceAmountFormat()): String {
    if (!v.isFinite()) return v.toString()
    val plain = BigDecimal.valueOf(v).stripTrailingZeros()
    // `scale() <= 0` is a whole number, and `toPlainString` on one of those can
    // still carry a trailing `.0`-shaped scale; go through the integer form so
    // "3" is "3".
    val text = if (plain.scale() <= 0) plain.toBigInteger().toString() else plain.toPlainString()
    return if (format == AmountFormat.COMMA) text.replace('.', ',') else text
}

/**
 * A typed amount as a number, or null if it is not one yet.
 *
 * The comma is the point. `KeyboardType.Decimal` shows whatever separator the
 * phone's locale uses, which across most of Europe is `,` — so "8,5" is what the
 * keyboard invites and `toDoubleOrNull` refuses, and the old elvis turned that
 * refusal into a silent target of 0 that no entry could ever meet. Half-typed
 * text still returns null; [HabitFormScreen] disables Save for it rather than
 * guessing, which is the same treatment the reminder time already gets.
 *
 * **A thousands group is refused, not guessed at.** A blanket comma-to-dot
 * replace read "10,000" as TEN, so a habit created with a goal of ten thousand
 * steps stored one of ten — a goal every day meets, on a habit that then looks
 * permanently complete and says nothing about it. The server cannot catch it
 * either: `parseHabit` coerces with `Number()` and bounds nothing above, so ten
 * is a perfectly valid target. It is genuinely ambiguous — "1,500" is fifteen
 * hundred to one reader and one and a half to another — and refusing is loud,
 * where guessing is silent and sometimes wrong by a factor of a thousand. A
 * non-zero integer part is required, so it fires on "10,000" and not on "0,255"
 * or ",255", where there are no thousands to separate.
 *
 * **[format] decides which spelling is a group, and it decides ONE input class
 * in both directions — say the second one out loud, because a first version of
 * this KDoc claimed it away.** Neither convention ever ACCEPTS its own group:
 * "8,5" and "8.5" are eight and a half under both, since a group is exactly
 * three digits, so for everything anyone ordinarily types the convention is
 * inert. Where it bites is a three-digit group, and there it cuts both ways.
 * Under [AmountFormat.COMMA], "10,000" is **ten** (ten and zero thousandths) —
 * the same answer `shared/public/ui/amount.js` gives for that input on a comma
 * account, and correct for a reader whose decimal point IS a comma.
 *
 * So the honest safety statement is conditional, not absolute. Under `auto` —
 * the `numberFormat` default and almost every account — the convention is
 * resolved from the same device the typing happens on, so the reading and the
 * typist agree by construction and a wrong guess is not reachable. What IS
 * reachable is the mismatch this client knowingly does not close: an account
 * that has explicitly chosen `point` on a comma-locale phone types "10,000"
 * meaning ten thousand, the web refuses it as ambiguous, and THIS reader takes
 * it as ten and stores it. That is a silent row out by a thousand, for that one
 * account class and that one input class — it is what a mirror would buy, and
 * it is more than "a better refusal message". Do not restate the older, flatter
 * claim that a wrong guess can only ever refuse: `android-native/CLAUDE.md` and
 * `docs/decisions/amounts.md` both carry the corrected version, and #157's own
 * open half is exactly this.
 *
 * `toDoubleOrNull` is generous in other ways nobody types into a goal box:
 * "1e3" is a thousand, "0x10" is sixteen, "Infinity" and "NaN" both parse, and
 * "-5" is negative. Digits, at most one separator, nothing else.
 *
 * Note what this is NOT. `shared/public/ui/amount.js` reads a day's AMOUNT in
 * the web app and this reads a habit's TARGET here (among other surfaces); they
 * are not mirrors and the root CLAUDE.md says why — a client mirrors a rule
 * only if it must work offline, and creating a habit is server-authoritative.
 * What is shared is the refusal, because the ambiguity is the same and so is
 * the cost of getting it wrong. Do not copy the web's `MAX_AMOUNT`/`MIN_AMOUNT`
 * bounds or its six-place quantisation here — the two deliberately part in
 * their DOMAIN and agree only about the FORM.
 */
internal fun parseAmount(text: String, format: AmountFormat = deviceAmountFormat()): Double? {
    val trimmed = text.trim()
    // A group that reads as a thousands separator, spelled the way [format]
    // says a group is spelled.
    if (isGroup(trimmed, format)) return null

    val decimal = trimmed.replace(',', '.')
    // Digits with at most one separator, and nothing else — no sign, no
    // exponent, no hex, no word.
    if (!DECIMAL.matches(decimal)) return null

    return decimal.toDoubleOrNull()?.takeIf { it.isFinite() && it >= 0 }
}

private fun isGroup(trimmed: String, format: AmountFormat): Boolean =
    when (format) {
        AmountFormat.POINT -> GROUP_POINT.matches(trimmed)
        AmountFormat.COMMA -> GROUP_COMMA.matches(trimmed)
    }

/**
 * "10,000" and "1,500", but not "0,255" or ",255" — a comma-separated group,
 * refused under [AmountFormat.POINT], where a comma reads as a thousands
 * separator. See [parseAmount].
 *
 * **ANCHORED and matched whole**, not the unanchored `containsMatchIn` this
 * used to be: the unanchored form is quadratic on a long run of digits
 * (`js/polynomial-redos`), which `GROUPS` in `shared/public/ui/amount.js`'s own
 * comment records as the reason for its own anchoring. This is not a narrowing
 * for the parser — everything the loose form caught and this does not
 * ("12,345,678", "10,000.5", "1,500 steps") has more than one separator or a
 * non-digit in it, so [DECIMAL] refuses it a line later regardless.
 *
 * The `0*` is load bearing: without it "01,000" is not a group here while
 * "01.000" IS a number to [DECIMAL], so a leading zero would buy a silent 1.
 */
private val GROUP_POINT = Regex("^0*[1-9]\\d*,\\d{3}$")

/**
 * "10.000" and "1.500", but not "0.255" or ".255" — a dot-separated group,
 * refused under [AmountFormat.COMMA], where a dot reads as a thousands
 * separator. See [GROUP_POINT] for why this is anchored and what the `0*`
 * guards against.
 */
private val GROUP_COMMA = Regex("^0*[1-9]\\d*\\.\\d{3}$")

/** Digits and at most one separator. */
private val DECIMAL = Regex("^(\\d+(\\.\\d*)?|\\.\\d+)$")

/**
 * Why [parseAmount] refused, for someone looking at the box.
 *
 * "Not a number" is true of "eight" and unhelpful for "10,000", which IS a
 * number and a perfectly reasonable thing to type — it is refused because it
 * is ambiguous, not because it is nonsense. A refusal the user cannot act on
 * is only half better than the silent ten it replaced.
 *
 * **The trigger is [parseAmount] itself, twice, and not [isGroup] — the box
 * was refused, and removing every grouper makes it an amount.** `isGroup`
 * ANCHORED (`^0*[1-9]\d*[,.]\d{3}$`) so it does not fire on "1,234,567" —
 * more than one group, past what a thousands separator alone can explain —
 * so a build that asked it directly told that input "Not a number", where the
 * PARSE is unaffected by the anchoring: `DECIMAL` still refuses it a line
 * later regardless of which guard named it. The actionable sentence is still
 * owed there, because taking every comma out of "1,234,567" DOES leave
 * something [parseAmount] accepts, and the sentence claims exactly that.
 * "10,000 steps" answers "Not a number" either way, correctly: taking the
 * commas out leaves "10000 steps", still not an amount, so the box may not
 * suggest something it would refuse.
 *
 * A literal, constant sentence per convention — never a template over the
 * user's input. `shared/test/amount.test.js` regexes the literal out of this
 * source; a template would lose that and the cross-client agreement test
 * would go green while saying nothing. Only the TRIGGER above decides which
 * of the two sentences is shown; neither sentence names what was typed.
 */
internal fun amountComplaint(text: String, format: AmountFormat = deviceAmountFormat()): String {
    val grouper = if (format == AmountFormat.COMMA) '.' else ','
    val plain = text.trim().filter { it != grouper }
    val ambiguous = parseAmount(text, format) == null && parseAmount(plain, format) != null
    return when {
        !ambiguous -> "Not a number"
        format == AmountFormat.COMMA ->
            "Type it without the thousands separator — 10000, not 10.000."
        else ->
            "Type it without the thousands separator — 10000, not 10,000."
    }
}
